### PR TITLE
[Fix #5165] don't allow invalid field names

### LINKED ISSUE #5165: Minimongo difference: it allows $ prefixed keys
If you try to use `$` prefixed key on the server, you will get an error. But Minimongo allows them. For example, for key `$each` you will get the following error:

```
$each is not valid for storage.
```

See reproduction (test `each-storage`) here: https://github.com/mitar/minimongo-diff/tree/split-tests

### PR DESCRIPTION
**Pull request by mutdmour**
#5165 

This fixes issues with minimongo allowing updates with invalid field names. I started out by looking at $addToSet as per ticket, and then expanded to check on other modifiers. I also checked on insert which was handling this issue. 

I ran all the tests on server MongoDB to confirm consistent behavior.

### CODE REVIEW

File: packages/minimongo/helpers.js [hunk 1]
```diff
@@ -43,3 +43,28 @@ isOperatorObject = function (valueSelector, inconsistentOK) {
...
+};
+isValidFieldName = function (key){
+      let match;
+      if (_.isString(key) && (match = key.match(/^\$|\.|\0/))) {
+        throw MinimongoError(`Key ${key} must not ${invalidCharMsg[match[0]]}`);
+      }
+}
+
+// checks if all field names in an object are valid
+hasValidFieldNames = function(doc){
```
Comment (benjamn) [inline packages/minimongo/helpers.js]: Let's call this `assertHasValidFieldNames`, and export it using ECMAScript syntax:
```js
export function assertHasValidFieldNames(doc) {
   ...
}
```
Then `import` it elsewhere:
```js
import { assertHasValidFieldNames } from "./helpers.js";
```

File: packages/minimongo/helpers.js [hunk 2]
```diff
@@ -43,3 +43,28 @@ isOperatorObject = function (valueSelector, inconsistentOK) {
...
+const invalidCharMsg = {
+  '.': "contain '.'",
+  '$': "start with '$'",
+  '\0': "contain null bytes",
+};
+isValidFieldName = function (key){
+      let match;
+      if (_.isString(key) && (match = key.match(/^\$|\.|\0/))) {
+        throw MinimongoError(`Key ${key} must not ${invalidCharMsg[match[0]]}`);
+      }
```
Comment (benjamn) [inline packages/minimongo/helpers.js]: Fix indentation (two spaces per tab) please.

File: packages/minimongo/helpers.js [hunk 3]
```diff
@@ -43,3 +43,28 @@ isOperatorObject = function (valueSelector, inconsistentOK) {
...
+
+// Make sure field names do not contain Mongo restricted
+// characters ('.', '$', '\0').
+// https://docs.mongodb.com/manual/reference/limits/#Restrictions-on-Field-Names
+const invalidCharMsg = {
+  '.': "contain '.'",
+  '$': "start with '$'",
+  '\0': "contain null bytes",
+};
+isValidFieldName = function (key){
```
Comment (benjamn) [inline packages/minimongo/helpers.js]: My comment about `assertHasValidFieldNames` applies here too (`assertIsValidFieldName`, `export`, `import`).

File: packages/minimongo/helpers.js (hunk 1 above)
Comment (mutdmour) [inline packages/minimongo/helpers.js]: Is the exporting/importing to help make it clearer where the function is coming from?

**Comment by mutdmour** [comment]
@benjamn not sure why this run test is failing, has nothing to do with minimongo.

**Comment by hwillson** [comment]
@mutdmour I've triggered a rebuild; sometimes that's all that is needed (for timeouts). We'll know shortly - thanks!

File: packages/minimongo/minimongo.js [hunk 4]
```diff
@@ -1,3 +1,5 @@
+import { assertHasValidFieldNames } from './validation.js';
```
Comment (benjamn) [inline packages/minimongo/minimongo.js]: Yep, this is a lot clearer to me!

**Comment by abernix** [comment]
This should be available to test in the release candidate for Meteor 1.4.4.2, currently `1.4.4.2-rc.1`:

    meteor update --release 1.4.4.2-rc.1

Please report any problems you encounter!

**Comment by brucejo75** [comment]
I just entered #8775, this seems like a pretty fundamental issue that was not caught.

Seems like the tests should have caught this...

**Pull request merged**